Generate nulls for parameters annotated as Nullable

berrueta opened the pull request:
This allows to write properties that take nullable parameters. This is particularly useful in Kotlin, where explicitly nullable types are very common (they serve similar purposes to the Optional type in Java 8+).

Rather than adding support for all the annotations that can indicate nullability (for example, these are the ones that the Kotlin compiler understands: https://github.com/JetBrains/kotlin/blob/master/core/descriptors.jvm/src/org/jetbrains/kotlin/load/java/JvmAnnotationNames.kt#L21 ), I've kept it to a minimum, but enough to have interoperability with Kotlin.

In the absence of annotations, null will not be generated, therefore this change is backwards compatible and will not break existing properties in Java or Kotlin.

pholser commented:
@berrueta Thanks for this! I'll review and get back with you soon!

pholser commented:
@berrueta A couple of questions for you:

1) It looks as though the change makes it so that, when deciding what set of generators could produce values for a given parameter, `GeneratorRepository` will include the `NullGenerator` in the set of matching generators if the parameter is marked with one of the recognized `@Nullable` annotations. I wonder if, instead, `NullGenerator` could be more like a decorator on the set of matching generators. When asked to `generate()`, the `NullGenerator` could produce `null` with some (configurable) probability `p`, and defer to the matching set with probability `1 - p`. Do you think such a design might prove even more useful for test writers?

2) What should happen in the event of `@Nullable` being discovered on a primitive parameter? Do we lean on compilers not to allow this?

3) Whatever design we land on, would you be willing to add some discussion to the Markdown docs? Feel free to add yourself as a contributor in the root POM also.

Cheers!

berrueta commented:
Hi @pholser ,

Thank you for your feedback. I intentionally aimed for a minimalistic change in this first iteration so I could receive some feedback. I like your suggestion to turn `NullGenerator` into a decorator. That would allow not only to configure the probability of nulls, but also to extend the nullability to other types like arrays and enums. I'm going to give it a try and I'll update the PR.

Do you have any suggestion on how to expose the configurable probability `p`? Perhaps a new annotation?

I'll confirm the behaviour in the case of primitive types, and add some documentation.

pholser commented:
@berrueta For configuring the probability of `null` generation on a parameter, I'm thinking either 1) a configuration annotation, 2) an attribute on the `@Property` annotation, 3) a separate annotation, maybe something like `com.pholser.junit.quickcheck.NullsAllows(probability = ...)`; along with what [jqwik](https://jqwik.net) does, for the `Gen` interface add something like its `injectNulls`.

berrueta commented:
Hi @pholser ,

I followed your suggestion and transformed `NullGenerator` into a decorator called `NullableGenerator`. I also added a check to make sure nulls are not generated for primitive types.

I'm not done yet. I still have to make the probability `p` configurable, add the documentation, etc., but I would appreciate if you could give me some feedback on the new design.

Thank you.

berrueta commented:
I've pushed a commit introducing a new `NullAllows` annotation (tentative name). I have a couple of design questions, please let me know what you think:

* By design, the new annotation is not sufficient by itself to indicate that the parameter is nullable. I didn't want to introduce yet another nullable annotation. Instead, I see it as a companion to the conventional nullable annotations like `javax.annotation.Nullable`. 

* Also by design, the new annotation is not a `GeneratorConfiguration`. I feel like the semantics are slightly different from the other generator configuration annotations. For example, it is not expected that the delegate generator (e.g., `IntegerGenerator`) will understand this annotation.

Any feedback would be welcome.

pholser commented on core/src/main/java/com/pholser/junit/quickcheck/internal/generator/GeneratorRepository.java:
```diff
@@ -59,6 +62,11 @@ a copy of this software and associated documentation files (the
 import static com.pholser.junit.quickcheck.internal.Reflection.*;
 
 public class GeneratorRepository implements Generators {
```
General note: I tend to favor static-importing names when doing so doesn't introduce confusion. For example, here I'd static-import `Collections.unmodifiableSet`, but not `Stream.of`.

pholser commented on core/src/main/java/com/pholser/junit/quickcheck/internal/generator/GeneratorRepository.java:
```diff
@@ -362,4 +374,16 @@ private boolean hasGeneratorsFor(ParameterTypeContext parameter) {
         return Types.forJavaLangReflectType(type);
     }
+
+    private static boolean isPrimitiveType(Type type) {
```
You may be able to avoid compiler warnings about raw types by phrasing as:

    return type instanceof Class<?>  && ((Class<?>) type).isPrimitive();

You can use the straight-wildcard form of a generic type in `instanceof` and casts without issue.

pholser commented on core/src/main/java/com/pholser/junit/quickcheck/internal/generator/NullAllows.java:
```diff
@@ -0,0 +1,21 @@
+ * to configure the probability of generating a null value.</p>
+ */
+@Target({ PARAMETER, FIELD, ANNOTATION_TYPE, TYPE_USE })
+@Retention(RUNTIME)
```
A couple of things here:

1) I think `NullsAllowed` or `NullAllowed` reads better for the name of this annotation.
2) What do you think about using only this annotation to mark nullable property parameters, instead of the others (`javax.annotation.Nullable`, etc.) in conjunction with this one?
3) 0.5 seems like a pretty high default. Could we make this 0.25, or 0.2, perhaps?

berrueta commented:
Hi @pholser ,

I've applied all the changes requested and added the documentation.

> What do you think about using only this annotation to mark nullable property parameters, instead of the others (javax.annotation.Nullable, etc.) in conjunction with this one?

Initially I preferred a design that made the new `NullAllowed` annotation orthogonal to `j.a.Nullable`, for the reasons I mentioned in a previous comment. However I have just changed my mind and made the `NullAllowed` annotation a sufficient condition to indicate that a parameter is nullable. The main reason for the change was the realisation that `org.jetbrains.annotations.Nullable` is not retained at runtime, which means that Kotlin users of junit-quickcheck must explicitly add at least one annotation if they want to generate null values.

Let me know if you have any other suggested change.

pholser commented:
@berrueta Thanks Diego! I'll have a look soon.

pholser commented:
@berrueta Thanks for the contribution -- this looks really good.
